addons/mail/static/src/messaging/component/composer/composer.js

```diff
@@ -321,11 +314,8 @@ class Composer extends Component {
 
     /**
      * @private
-     * @param {CustomEvent} ev
      */
-    _onTextInputKeydownEnter(ev) {
-        ev.stopPropagation();
-        // TODO SEB this is the same code as _onClickSend
+    _onSend(ev) {
```

Comment:
This method doesn't really make sense to be honest. I prefer something like this:

```js

_canPostMessage() {
   return (
      !this.composer.textInputContent &&
      this.composer.attachments.length === 0
   );
}

_onClickSend(ev) {
   if (!this._canPostMessage()) {
      return;
   }
   return this._postMessage();
}

_onTextInputKeydownEnter(ev) {
   if (!this._canPostMessage()) {
      return;
   }
   return this._postMessage();
}
```

Ideally we shouldn't have `_postMessage` here, but that's not easy to change and there's actually a task to do some "spying" of entity changes in a similar way as component hints:
https://www.odoo.com/web#id=2252858&action=3206&active_id=1519&model=project.task&view_type=form&cids=1&menu_id=4720
(the task was created with concern to deletion of entity/components, but that's exactly the same underlying problem and required solution)

Reply:
But you can keep `_postMessage` in the meantime ;)

Reply:
So ?

Reply:
> So ?

Something's confusing with my comment? If so, what did you not understand?

Reply:
Should I just apply the changes you suggest without bother of the location of ```_postMessage``` ?

Reply:
You can keep `_postMessage`, in my suggestion I have omitted sections (`// Private`, `// Handlers`) and docstrings, but obviously you should have them + alphabetical order of methods.

Reply:
Does it make sense to have two methods doing the exact same thing ? ....

Reply:
I don't mind having same inner code for `_onClickSend` and `_onTextInputKeydownEnter `. They are small and easy to understand. Having a separate function just to share 3 LOCs adds a weird and unintuitive indirection IMO (I can't find a good way to name it, so this is an indication that we shouldn't isolate in a function), so I'm good with small duplication of code (I think this is the most maintainable solution).

Reply:
Is there a problem with an handler callable from 2 differents locations ? Couldn't those two methods be the same ? I don't like at all the idea of code duplication even for 3 LOC ....

Reply:
No, don't share same handler, if we have a crash we don't know from which event it happens, hence one event => one handler.

I don't mind code duplication right now, especially when this is small like this. I really think that if we do something to avoid this code duplication, this would be more harmful than just duplicate the code.

Reply:
Done but there was a mistake in ```_canPostMessage```. As it was, the method returned false when a message could possibly be posted.